### PR TITLE
Create new error E0742 and add long error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #61137.

Creates E0742 error code and add its long error explanation.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
cc @Dylan-DPC :)

File: src/librustc_resolve/error_codes.rs [hunk 1]
```diff
@@ -1720,6 +1720,44 @@ struct Foo<X = Box<Self>> {
 ```
 "##,
 
+E0740: r##"
+A visibility has been restricted to a module which isn't an ancestor of the
```
Comment (Dylan-DPC) [inline src/librustc_resolve/error_codes.rs]:
```suggestion
Visibility is restricted to a module which isn't an ancestor of the
```

**Review by Dylan-DPC-zz — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
I changed the error code to fix tests.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/65769, @bors r-

**Comment by GuillaumeGomez** [comment]
Erf, someone already created E0741... :'(

**Comment by Dylan-DPC** [comment]
We need Mutex<Error Codes> ;)

**Comment by GuillaumeGomez** [comment]
Indeed haha!

**Comment by GuillaumeGomez** [comment]
And to avoid having to update the error code once again:

@bors: p=1

**Pull request merged**